src/psyclone/psyir/transformations/arrayassignment2loops_trans.py

```diff
@@ -368,6 +367,34 @@ def validate(self, node, options=None):
                 # pylint: disable=cell-var-from-loop
                 raise TransformationError(LazyString(
                     lambda: f"{message} In:\n{node.debug_string()}"))
+            # We must understand any array accesses on the RHS. This means all
+            # array indices must be known (because otherwise they themselves
+            # might be arrays, e.g.
+            #      a(:) = b(c) where `c` can be a scalar or an array.
+            if isinstance(reference, ArrayMixin):
+                # Check the index expressions
+                range_count = len([x for x in accessor.indices
+                                   if isinstance(x, Range)])
+                if range_count != num_of_ranges:
+                    # If the number of ranges in this access is not the same as
+                    # on the LHS then we may or may not have a scalar.
+                    for exprn in reference.indices:
+                        if not isinstance(exprn, Reference):
+                            continue
```

Comment:
I am not sure that's right, for example in the new "test_apply_indirect_indexing" if in
`ishtSi(5:8,jf) = ishtSi( iwewe,jf )`
if iwewe was not a Reference and instead a codeblock:
`ishtSi(5:8,jf) = ishtSi( (/ jpwe,jpea,jpwe,jpea /) , jf )`
it should not validate.

Similarly, if it was a function call that is not elemental but we don't know if it returns a scalar or an array.

By contrast, I think we will probably do the right thing for expressions, e.g.:
`ishtSi(5:8,jf) = ishtSi( iwewe + 1, jf )`
as expr.datatype with bubble up the UnresolvedType if any part of the expression is unresolved.

Please add tests for all these.

Reply:
Contrary to what you expected, we already refused to validate your first two examples. However, for the third one we haven't yet implemented support for finding the datatype when there are expressions in the array indices (#1799). I've added tests for all of these along with an associated TODO for the last one.